sdk/messaging/azservicebus/messageBatch.go

```diff
@@ -56,48 +85,34 @@ func (mb *MessageBatch) Add(m SendableMessage) (bool, error) {
 
 	bin, err := msg.MarshalBinary()
 	if err != nil {
-		return false, err
+		return err
 	}
 
-	if mb.Size()+len(bin) > int(mb.maxBytes) {
-		return false, nil
+	if int(mb.NumBytes())+len(bin) > int(mb.maxBytes) {
+		return ErrMessageTooLarge
 	}
 
-	mb.size += len(bin)
+	mb.size += int32(len(bin))
 
 	if len(mb.marshaledMessages) == 0 {
 		// first message, store it since we need to copy attributes from it
 		// when we send the overall batch message.
-		amqpMessage := m.toAMQPMessage()
-
-		// we don't need to hold onto this since it'll get encoded
-		// into our marshaledMessages. We just want the metadata.
-		amqpMessage.Data = nil
-		mb.batchEnvelope = amqpMessage
+		mb.batchEnvelope = createBatchEnvelope(msg)
 	}
 
 	mb.marshaledMessages = append(mb.marshaledMessages, bin)
-	return true, nil
-}
-
-// Size is the number of bytes in the message batch
-func (mb *MessageBatch) Size() int {
-	// calculated data size + batch message wrapper + data wrapper portions of the message
-	return mb.size + batchMessageWrapperSize + (len(mb.marshaledMessages) * 5)
-}
-
-// Len returns the # of messages in the batch.
-func (mb *MessageBatch) Len() int {
-	return len(mb.marshaledMessages)
+	return nil
 }
 
-// toAMQPMessage converts this batch into a sendable *amqp.Message
-// NOTE: not idempotent!
-func (mb *MessageBatch) toAMQPMessage() *amqp.Message {
-	mb.batchEnvelope.Data = make([][]byte, len(mb.marshaledMessages))
-	mb.batchEnvelope.Format = batchMessageFormat
+// createBatchEnvelope makes a copy of the properties of the message, minus any
+// payload fields (like Data, Value or (eventually) Sequence). The data field will be
+// filled in with all the messages when the batch is completed.
+func createBatchEnvelope(am *amqp.Message) *amqp.Message {
+	newAMQPMessage := *am
 
-	copy(mb.batchEnvelope.Data, mb.marshaledMessages)
+	// TODO: we appear to be missing the .Sequence section
```

Comment:
Author of the above comment seems to have additional insight here :suspect:

Reply:
Ha, I wish that guy would work here. ;) 

Removed the comment and filed https://github.com/Azure/go-amqp/issues/94, as it's something we need to discuss for full AMQP support. It looks like (based on the TODO comment in the go-amqp source) that they're actually squishing sequence into [][]Data so technically my comment is "incorrect".